app/Jobs/FetchNativeBalances.php

```diff
@@ -6,52 +6,102 @@
+use Illuminate\Support\Str;
 
 class FetchNativeBalances implements ShouldBeUnique, ShouldQueue
 {
-    use Dispatchable, InteractsWithQueue, Queueable, RecoversFromProviderErrors, SerializesModels, WithWeb3DataProvider;
+    use Dispatchable, InteractsWithQueue, Queueable, RecoversFromProviderErrors, WithWeb3DataProvider;
 
+    /**
+     * @var Collection<int, Wallet>
+     */
+    public Collection $wallets;
+
+    /**
+     * @param  Collection<int, Wallet>  $wallets
+     */
     public function __construct(
-        public Wallet $wallet,
+        Collection|Wallet $wallets,
         public Network $network,
     ) {
+        $this->wallets = $wallets instanceof Wallet ? collect([$wallets]) : $wallets;
     }
 
     public function handle(): void
     {
+        Log::info('Processing FetchNativeBalances job', [
+            'network_id' => $this->network->id,
+            'wallet_ids' => $this->wallets->pluck('id')->toArray(),
+        ]);
+
         $nativeToken = $this->network->nativeToken()->firstOrFail();
 
-        $provider = $this->getWeb3DataProvider();
-        $balance = $provider->getNativeBalance($this->wallet, $this->network);
+        $addresses = $this->wallets->pluck('address')->toArray();
+
+        $balances = Moralis::getNativeBalances($addresses, $this->network);
+
+        $walletsToUpdate = collect();
+
+        $balancesToInsert = $balances->map(function ($walletBalance) use ($nativeToken, $walletsToUpdate) {
+            $address = Str::lower($walletBalance->address);
+
+            $wallet = $this->wallets->first(fn ($wallet) => $address === Str::lower($wallet->address));
+
+            $walletsToUpdate->push($wallet);
 
-        DB::transaction(function () use ($balance, $nativeToken) {
-            DB::table('balances')->upsert([
-                'wallet_id' => $this->wallet->id,
+            return [
+                'wallet_id' => $wallet->id,
                 'token_id' => $nativeToken->id,
-                'balance' => $balance,
+                'balance' => $walletBalance->balance,
                 'created_at' => now(),
                 'updated_at' => now(),
-            ], ['wallet_id', 'token_id'], ['balance', 'updated_at']);
+            ];
+        });
+
+        if ($balancesToInsert->isEmpty()) {
+            return;
+        }
+
+        Log::info('Updating native balances', [
+            'network_id' => $this->network->id,
+            'data' => $balancesToInsert->map(fn ($balance) => collect($balance)->only(['wallet_id', 'balance'])),
+        ]);
+
+        DB::transaction(function () use ($balancesToInsert) {
+            Balance::query()->upsert(
+                $balancesToInsert->toArray(),
+                ['wallet_id', 'token_id'],
+                ['balance', 'updated_at']
+            );
+        });
+
+        $walletsToUpdate->map(function ($wallet) {
+            $wallet->extra_attributes->set('native_balances_fetched_at', Carbon::now());
+            $wallet->save();
         });
```

Comment:
Is there a reason this isn't a part of the database transaction?

Reply:
Also, this is precisely why I hate these JSON columns, they just make things difficult, with no real benefits 🤷  I'd be totally okay if you just added the new timestamp column and then you could update in one query 🙈

Reply:
I didn't put too much though on moving `native_balances_fetched_at` into transaction, however, I should, if balance update fails then  `native_balances_fetched_at` should have `null`/old value, so the job can pick it next time.

Reply:
I checked many places to find at what point it gets `[]` set but couldn't find it, if I could fix that I'd set `native_balances_fetched_at` in one query..